Implement the get(name) function in Cookie Store API

RupinMittal opened the pull request:
#### 9ae242a9a2f095eed7c17d453874a733bd57ce92
<pre>
Implement the get(name) function in Cookie Store API
<a href="https://bugs.webkit.org/show_bug.cgi?id=259071">https://bugs.webkit.org/show_bug.cgi?id=259071</a>

Reviewed by Chris Dumez.

CookieStore (on the WebProcess side), will make an IPC to the
Network Process (via WebCookieJar) and will pass the relevant
CookieStoreGetOptions through the IPC. The Network process will
retrieve the stored cookies, filter them using the provided
options and will asynchronously resolve the promise with a
vector of the filtered cookies. In the get(name) function, the
promise is resolved with only the first cookie in the vector
(if there are any at all).

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/cookie-store/CookieListItem.idl:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::create):
(WebCore::CookieStore::CookieStore):
(WebCore::CookieStore::get):
(WebCore::CookieStore::activeDOMObjectName const):
(WebCore::CookieStore::scriptExecutionContext const):
* Source/WebCore/Modules/cookie-store/CookieStore.h:
* Source/WebCore/Modules/cookie-store/CookieStore.idl:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::cookieStore):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::vectorOfCookiesForSession const):
(WebCore::NetworkStorageSession::vectorOfCookiesForDOM const):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::cookiesForDOMAsync):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::getCookiesAsync const):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: <a href="https://commits.webkit.org/265963@main">https://commits.webkit.org/265963@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e8e36d62db2d136a809c00e0ccb803a21ea0501

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12420 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12755 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13066 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14163 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/11939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12479 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15181 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12773 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14615 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12584 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13331 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10499 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14585 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10614 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11244 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18343 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11697 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11409 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14596 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/11903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9825 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11136 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3050 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15465 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/11758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -38,18 +45,45 @@ namespace WebCore {
+    auto cookieStoreGetoptions = CookieStoreGetOptions { name, { } };
+    auto completionHandler = [promise = WTFMove(promise)] (Vector<Cookie>&& cookies) {
+        if (cookies.isEmpty()) {
+            promise->resolveWithJSValue(JSC::jsNull());
```
I tried other ways of writing this line, but none of them seemed to work, so I kept this.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -26,10 +26,17 @@
+#include "JSCookieListItem.h"
 #include "JSDOMPromiseDeferred.h"
+#include "ScriptExecutionContext.h"
+#include <wtf/Forward.h>
```
Seems odd to include Forward.h in a cpp.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -38,18 +45,45 @@ namespace WebCore {
+    auto& url = context->url();
+    auto& document = *downcast<Document>(context);
+    auto& cookieJar = document.page()->cookieJar();
+    auto cookieStoreGetoptions = CookieStoreGetOptions { name, { } };
```
Let's call this `options`:
`CookieStoreGetOptions options { name, { } };`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -38,18 +45,45 @@ namespace WebCore {
 WTF_MAKE_ISO_ALLOCATED_IMPL(CookieStore);
 
-Ref<CookieStore> CookieStore::create()
+Ref<CookieStore> CookieStore::create(ScriptExecutionContext& context)
```
This should take a `Document&`, otherwise your downcast<Document> below looks unsafe.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -38,18 +45,45 @@ namespace WebCore {
 }
 
-CookieStore::CookieStore() = default;
+CookieStore::CookieStore(ScriptExecutionContext& context)
```
Should take a Document& for now.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -38,18 +45,45 @@ namespace WebCore {
+    auto& document = *downcast<Document>(context);
+    auto& cookieJar = document.page()->cookieJar();
+    auto cookieStoreGetoptions = CookieStoreGetOptions { name, { } };
+    auto completionHandler = [promise = WTFMove(promise)] (Vector<Cookie>&& cookies) {
```
Most of this logic should be moved to the get() overload which takes in an Options struct. Then this function would just creates the Options struct and call the other get() function with it.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -38,18 +45,45 @@ namespace WebCore {
+
+        auto& cookie = cookies[0];
+        CookieListItem cookieListItem = { cookie.name, cookie.value };
+        promise->resolve<IDLDictionary<CookieListItem>>(cookieListItem);
```
`promise->resolve<IDLDictionary<CookieListItem>>(CookieListItem { WTFMove(cookie.name), WTFMove(cookie.value) });`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.h:
```diff
@@ -36,11 +37,12 @@ struct CookieInit;
 struct CookieStoreDeleteOptions;
 struct CookieStoreGetOptions;
 class DeferredPromise;
+class ScriptExecutionContext;
```
I doubt this is needed.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.h:
```diff
@@ -59,7 +61,10 @@ class CookieStore final : public RefCounted<CookieStore>, public EventTarget {
 
 private:
-    CookieStore();
+    CookieStore(ScriptExecutionContext&);
```
explicit

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -186,4 +189,8 @@ void CookieJar::deleteCookie(const Document&, const URL& url, const String& cook
 }
 
+void CookieJar::getCookiesAsync(Document&, const URL&, const CookieStoreGetOptions&, CompletionHandler<void(Vector<Cookie>&&)>&&) const
+{
```
This should call the completion handler or it would crash in debug.

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2807,7 +2807,7 @@ WebCoreOpaqueRoot root(LocalDOMWindow* window)
 {
     if (!m_cookieStore)
-        m_cookieStore = CookieStore::create();
+        m_cookieStore = CookieStore::create(*(this->document()));
```
`m_cookieStore = CookieStore::create(document());`

Nothing guarantees that `document()` returns non-null so the CookieStore Factory should take in a `Document*` not a `Document&`

cdumez commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -390,11 +391,49 @@ - (CFURLRef)_cfurl;
+        }
+
+        auto convertedCookie = Cookie(cookie);
+        if (convertedCookie.name != cookieStoreGetOptions.name)
```
It is too wasteful to be doing this check *after* constructing the WebCore::Cookie object. You should do the check on the platform cookie type instead.

cdumez replied:
Also, cookieStoreGetOptions.name can be null, in which case we want to return ALL cookies. The current check is wrong.

RupinMittal commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_arguments.https.any-expected.txt:
```diff
@@ -6,10 +6,6 @@ FAIL cookieStore.get with name in options promise_test: Unhandled rejection with
-    [native code]
-}" ("TypeError")
+FAIL cookieStore.get with invalid url path in options assert_unreached: Should have rejected: undefined Reached unreachable code
+FAIL cookieStore.get with invalid url host in options assert_unreached: Should have rejected: undefined Reached unreachable code
```
I'm not sure that this change should have happened, but I put it here to get a second pair of eyes on it.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -26,35 +26,78 @@
 {
-    promise->reject(NotSupportedError);
+    CookieStoreGetOptions options { name, { } };
+    get(WTFMove(options), WTFMove(promise));
```
`get(CookieStoreGetOptions { WTFMove(name), { } }, WTFMove(promise));`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -26,35 +26,78 @@
 CookieStore::~CookieStore() = default;
 
-void CookieStore::get(const String&, Ref<DeferredPromise>&& promise)
+void CookieStore::get(const String& name, Ref<DeferredPromise>&& promise)
```
`String&&`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -26,35 +26,78 @@
+    }
+
+    if (origin->isOpaque()) {
+        promise->reject(SecurityError);
```
reject(Exception { SecurityError, "The origin is opaque"_s });

cdumez commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -390,11 +391,48 @@ - (CFURLRef)_cfurl;
+            if (includeSecureCookies == IncludeSecureCookies::No)
+                continue;
+        }
+        if (name && ![[cookie name] isEqualToString:name])
```
Still wrong:
```
inline String::operator NSString *() const
{
    if (!m_impl)
        return @"";
    return *m_impl;
}
```

so if `options.name` is a null string, `name` will be an *empty* NSString (not `nil`) and you'll return no cookies instead of all of them.

cdumez commented on Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp:
```diff
@@ -84,6 +87,12 @@ static std::pair<String, bool> cookiesForSession(const NetworkStorageSession& se
 
+Vector<Cookie> vectorOfCookiesForSession(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, IncludeHTTPOnlyOrNot includeHTTPOnly, IncludeSecureCookies includeSecureCookies, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, cons ...
+{
+    // FIXME: Implement for the Cookie Store API
```
Missing period at the end of the comment.

cdumez commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -637,6 +639,12 @@ std::pair<String, bool> NetworkStorageSession::cookiesForDOM(const URL& firstPar
 
+Vector<Cookie> vectorOfCookiesForSession(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, IncludeHTTPOnlyOrNot includeHTTPOnly, IncludeSecureCookies includeSecureCookies, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, cons ...
+{
+    // FIXME: Implement for the Cookie Store API
```
Missing period.

cdumez commented on Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp:
```diff
@@ -84,6 +87,12 @@ static std::pair<String, bool> cookiesForSession(const NetworkStorageSession& se
     return { cookies.toString(), didAccessSecureCookies };
 }
 
+Vector<Cookie> vectorOfCookiesForSession(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, IncludeHTTPOnlyOrNot includeHTTPOnly, IncludeSecureCookies includeSecureCookies, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, cons ...
```
missing `NetworkStorageSession::` ?

cdumez commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -637,6 +639,12 @@ std::pair<String, bool> NetworkStorageSession::cookiesForDOM(const URL& firstPar
     return cookiesForSession(*this, firstParty, url, sameSiteInfo, frameID, pageID, false, includeSecureCookies, applyTrackingPrevention, relaxThirdPartyCookieBlocking);
 }
 
+Vector<Cookie> vectorOfCookiesForSession(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, IncludeHTTPOnlyOrNot includeHTTPOnly, IncludeSecureCookies includeSecureCookies, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, cons ...
```
missing `NetworkStorageSession::` ?

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -41,6 +43,7 @@
 #include <WebCore/Page.h>
 #include <WebCore/Settings.h>
 #include <WebCore/StorageSessionProvider.h>
+#include <wtf/Forward.h>
```
Why is this needed?

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -26,35 +26,77 @@
+        return;
+    }
+
+    auto& url = context->url();
```
Could you please move this to the next line and call document.url() instead? It would avoid a virtual function call

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -26,35 +26,77 @@
+
+    auto& url = context->url();
+    auto& document = *downcast<Document>(context);
+    auto& cookieJar = document.page()->cookieJar();
```
The page needs null checking.

cdumez commented on Source/WebCore/platform/network/NetworkStorageSession.h:
```diff
@@ -171,6 +173,7 @@ class NetworkStorageSession : public CanMakeWeakPtr<NetworkStorageSession> {
     WEBCORE_EXPORT bool getRawCookies(const URL& firstParty, const SameSiteInfo&, const URL&, std::optional<FrameIdentifier>, std::optional<PageIdentifier>, ApplyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking, Vector<Cookie>&) const;
     WEBCORE_EXPORT void getHostnamesWithCookies(HashSet<String>& hostnames);
     WEBCORE_EXPORT std::pair<String, bool> cookiesForDOM(const URL& firstParty, const SameSiteInfo&, const URL&, std::optional<FrameIdentifier>, std::optional<PageIdentifier>, IncludeSecureCookies, ApplyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking) const;
+    WEBCORE_EXPORT Vector<Cookie> vectorOfCookiesForDOM(const URL& firstParty, const SameSiteInfo&, const URL&, std::optional<FrameIdentifier>, std::optional<PageIdentifier>, IncludeSecureCookies, ApplyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking, const CookieStoreGetOptions&) const;
```
Could we rename this to `cookiesForDOMAsVector()`. I think it would match better how we often name things in WebKit.

cdumez commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -390,11 +391,48 @@ - (CFURLRef)_cfurl;
+            if (includeSecureCookies == IncludeSecureCookies::No)
+                continue;
+        }
+        if (!!options.name && ![[cookie name] isEqualToString:name])
```
I don't think this addressed my previous review comment? It still doesn't seem to deal with null Strings correctly.

cdumez replied:
Actually, I misread, you're null-checking `options.name` now so it should work. I'd recommend using `options.name.isNull()` for clarity.

cdumez commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -390,11 +391,48 @@ - (CFURLRef)_cfurl;
 
+Vector<Cookie> NetworkStorageSession::vectorOfCookiesForDOM(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, IncludeSecureCookies includeSecureCookies, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, const CookieStoreGetOpt ...
+{
+    return vectorOfCookiesForSession(firstParty, sameSiteInfo, url, frameID, pageID, DoNotIncludeHTTPOnly, includeSecureCookies, applyTrackingPrevention, shouldRelaxThirdPartyCookieBlocking, options);
```
Similar concern about the naming, I'd use a `AsVector` suffix instead.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -26,35 +26,83 @@
+    }
+
+    auto& document = *downcast<Document>(context);
+    auto& url = document.url();
```
nit: Ideally we'd move this after the page null check too to avoid doing it when the page is null.

cdumez commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -390,11 +391,48 @@ - (CFURLRef)_cfurl;
+            continue;
+        if (!includeHTTPOnly && [cookie isHTTPOnly])
+            continue;
+        if ([cookie isSecure]) {
```
nit:
```
if ([cookie isSecure] && includeSecureCookies == IncludeSecureCookies::No)
    continue;
```

would look nicer IMO.

cdumez reviewed (approved):
LGTM with nits, assuming the bots are happy.

RupinMittal committed:
Implement the get(name) function in Cookie Store API
https://bugs.webkit.org/show_bug.cgi?id=259071

Reviewed by Chris Dumez.

CookieStore (on the WebProcess side), will make an IPC to the
Network Process (via WebCookieJar) and will pass the relevant
CookieStoreGetOptions through the IPC. The Network process will
retrieve the stored cookies, filter them using the provided
options and will asynchronously resolve the promise with a
vector of the filtered cookies. In the get(name) function, the
promise is resolved with only the first cookie in the vector
(if there are any at all).

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/cookie-store/CookieListItem.idl:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::create):
(WebCore::CookieStore::CookieStore):
(WebCore::CookieStore::get):
(WebCore::CookieStore::activeDOMObjectName const):
(WebCore::CookieStore::scriptExecutionContext const):
* Source/WebCore/Modules/cookie-store/CookieStore.h:
* Source/WebCore/Modules/cookie-store/CookieStore.idl:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::cookieStore):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::vectorOfCookiesForSession const):
(WebCore::NetworkStorageSession::vectorOfCookiesForDOM const):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::cookiesForDOMAsync):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::getCookiesAsync const):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: https://commits.webkit.org/265963@main
